src/libstd/sys/unix/mutex.rs

```diff
@@ -30,6 +30,20 @@ impl Mutex {
         Mutex { inner: UnsafeCell::new(libc::PTHREAD_MUTEX_INITIALIZER) }
     }
     #[inline]
+    pub unsafe fn init(&mut self) {
+        // We need to set the mutex type to PTHREAD_MUTEX_NORMAL to avoid
+        // undefined behavior when re-locking in the same thread.
```

Comment:
Could this expand a bit and link to the lock elision issue and perhaps a link to the standard as well? Something along the lines of the default being DEFAULT which has recursive locking as undefined behavior, but we need to guarantee somehow that a recursive lock never returns (either deadlock or panic)